pkg/registry/registrydeployer.go

```diff
@@ -0,0 +1,139 @@
+	interface_client "github.com/operator-framework/operator-marketplace/pkg/client"
+	"github.com/operator-framework/operator-marketplace/pkg/datastore"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// NewRegistryDeployer returns a new GrpcBuilder.
+func NewRegistryDeployer(log *logrus.Entry, reader datastore.Reader, client interface_client.Client) RegistryDeployer {
+	return RegistryDeployer{
+		log:    log,
+		reader: reader,
+		client: client,
+	}
+}
+
+// RegistryDeployer creates the resources required to deploy a registry and enable operators
+// on a cluster managed by OLM.
+type RegistryDeployer struct {
+	log    *logrus.Entry
+	reader datastore.Reader
+	client interface_client.Client
+}
+
+// CreateRegistryResources ensures a CatalogSource exists with all the
+// resources it requires.
+func (r *RegistryDeployer) CreateRegistryResources(key types.NamespacedName, displayName, publisher, targetNamespace, packages string, cscLabels map[string]string) error {
+	// Ensure that the packages in the spec are available in the datastore
+	err := r.checkPackages(v1.GetValidPackageSliceFromString(packages))
+	if err != nil {
+		return err
+	}
+
+	// Ensure that a registry deployment is available
+	registry := NewRegistry(r.log, r.client, r.reader, key, packages, RegistryServerImage)
+	err = registry.Ensure()
+	if err != nil {
+		return err
+	}
+
+	// Check if the CatalogSource already exists
+	catalogSourceGet := new(CatalogSourceBuilder).WithTypeMeta().CatalogSource()
+	err = r.client.Get(context.TODO(), key, catalogSourceGet)
+
+	// Update the CatalogSource if it exists else create one.
+	if err == nil {
+		catalogSourceGet.Spec.Address = registry.GetAddress()
+		r.log.Infof("Updating CatalogSource %s", catalogSourceGet.Name)
+		err = r.client.Update(context.TODO(), catalogSourceGet)
+		if err != nil {
+			r.log.Errorf("Failed to update CatalogSource : %v", err)
+			return err
+		}
+		r.log.Infof("Updated CatalogSource %s", catalogSourceGet.Name)
+	} else {
+		// Create the CatalogSource structure
+		catalogSource := newCatalogSource(cscLabels, key, displayName, publisher, targetNamespace, registry.GetAddress())
+		r.log.Infof("Creating CatalogSource %s", catalogSource.Name)
+		err = r.client.Create(context.TODO(), catalogSource)
+		if err != nil && !errors.IsAlreadyExists(err) {
+			r.log.Errorf("Failed to create CatalogSource : %v", err)
+			return err
+		}
+		r.log.Infof("Created CatalogSource %s", catalogSource.Name)
+	}
+
+	return nil
+}
+
+// EnsurePackagesInStatus makes sure that the csc's status.PackageRepositioryVersions
+// field is updated at the end of the configuring phase if successful. It iterates
+// over the list of packages and creates a new map of PackageName:Version for each
+// package in the spec.
+func (r *RegistryDeployer) EnsurePackagesInStatus(csc *marketplace.CatalogSourceConfig) {
```

Comment:
This should not be in the `CatalogSourceConfig` package

Reply:
In this commit the `OperatorSource` configuring reconciler is still creating a `catalogsourceconfig` so this had to go in the common package since it was being used by both. In the next commit since `OperatorSource` is no longer using this method, we decided to move it back.

Reply:
This sort of code churn is not advisable and makes it confusing. You should try to leave it where it is then.

Reply:
But that would mean duplicating it so that it's available in the operatorsource package

Reply:
Then it is pointing at that your abstractions are incorrect. The registry package should not have any `CatalogSourceConfig` specific code.

Reply:
Ok good point. Will revisit this